test/integration/sync-api-test.js

```diff
@@ -0,0 +1,307 @@
+        // handle them correctly.
+        getSignedInUserProfile: () => Promise.resolve({
+          displayName: undefined,
+          avatar: "https://firefoxusercontent.com/00000000000000000000000000000000",
+        }),
+        hasLocalSession: () => Promise.resolve(true),
+      };
+
+      await UIState.refresh();
+    })().then(arguments[0]);
+    `);
+  };
+
+  // Mock out the state where the user is logged out or has never logged in.
+  const setNotConfigured = async () => {
+    await webext.inChrome();
+    await driver.executeAsyncScript(`(async () => {
+      const testTime = new Date("January 1, 2019");
+      ChromeUtils.import("resource://gre/modules/Services.jsm");
+      Services.prefs.setCharPref("services.sync.lastSync", testTime);
+      UIState._internal.syncing = false;
+
+      UIState._internal.fxAccounts = {
+        getSignedInUser: () => Promise.resolve(null),
+        getSignedInUserProfile: () => Promise.resolve(null),
+      };
+
+      await UIState.refresh();
+    })().then(arguments[0]);
+    `);
+  };
+
+  // Mock out the state where the user is logged in, but the password was
+  // changed on another device.
+  const setLoginFailed = async () => {
+    await webext.inChrome();
+    await driver.executeAsyncScript(`(async () => {
+      const testTime = new Date("January 1, 2019");
+      ChromeUtils.import("resource://gre/modules/Services.jsm");
+      Services.prefs.setCharPref("services.sync.lastSync", testTime);
+      UIState._internal.syncing = false;
+
+      Services.prefs.setStringPref("services.sync.username", "foo@example.com");
+      UIState._internal.fxAccounts = {
+        getSignedInUser: () => Promise.resolve(null),
+      };
+
+      await UIState.refresh();
+    })().then(arguments[0]);
+    `);
+  };
+
+  // Mock out the state where the user is logged in, but the email is
+  // unverified.
+  const setNotVerified = async () => {
+    await webext.inChrome();
+    await driver.executeAsyncScript(`(async () => {
+      const testTime = new Date("January 1, 2019");
+      ChromeUtils.import("resource://gre/modules/Services.jsm");
+      Services.prefs.setCharPref("services.sync.lastSync", testTime);
+      UIState._internal.syncing = false;
+
+      UIState._internal.fxAccounts = {
+        getSignedInUser: () => Promise.resolve({
+          verified: false,
+          email: "foo@example.com",
+        }),
+        getSignedInUserProfile: () => Promise.resolve({
+          displayName: undefined,
+          avatar: "https://firefoxusercontent.com/00000000000000000000000000000000",
+        }),
+        hasLocalSession: () => Promise.resolve(true),
+      };
+
+      await UIState.refresh();
+    })().then(arguments[0]);
+    `);
+  };
+
+  const MockUserProfileInfo = {
+    loggedIn: {
+      avatar: "https://firefoxusercontent.com/00000000000000000000000000000000",
+      id: null, // TODO: why is the id null? am I setting this up wrong?
```

Comment:
Think I've got a fix for this: need to borrow some machinery from an FxA-specific test, https://searchfox.org/mozilla-central/source/services/fxaccounts/tests/xpcshell/test_accounts_device_registration.js

Reply:
This turns out to be [a lot of code](https://searchfox.org/mozilla-central/source/services/fxaccounts/tests/xpcshell/test_accounts_device_registration.js#94-128) to mock out, and I'm not quite sure it'll work. I'm inclined to leave FxA uid verification testing to a followup ticket. Thoughts?

Reply:
eoger in #sync on Slack suggested I just use some form of dependency injection, as in https://searchfox.org/mozilla-central/rev/dac799c9f4e9f5f05c1071cba94f2522aa31f7eb/services/fxaccounts/FxAccountsProfileClient.jsm#35. I guess I could expose an fxa injector method in the API, like `testSetFxAccounts()`, since I have no idea how the WebExtensions framework code starts up the API itself...but that seems gross.

Reply:
The problem, it turns out, was that I wasn't using the FxAccounts instance available on UIState._internal. I was loading FxAccounts.jsm.